src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -564,6 +580,27 @@ private TypeGenerationSpec ParseTypeGenerationSpec(in TypeToGenerate typeToGener
                 }
             }
 
+            private bool IsMemoryType(ITypeSymbol type, [NotNullWhen(true)] out INamedTypeSymbol? namedType, out CollectionType collectionType)
+            {
+                if (type.GetCompatibleGenericBaseType(_knownSymbols.MemoryType) is INamedTypeSymbol memoryType)
+                {
+                    namedType = memoryType;
+                    collectionType = CollectionType.MemoryOfT;
+                    return true;
+                }
+
+                if (type.GetCompatibleGenericBaseType(_knownSymbols.ReadOnlyMemoryType) is INamedTypeSymbol readOnlyMemoryType)
```

Comment:
Following up on what we discussed yesterday, I think you can actually replace this method with `SymbolEqualityComparer.Equals` as long you compare it with the type's `OriginalDefinition` symbol:

https://github.com/dotnet/runtime/blob/d59af2cf097acb100ad6a9cba652be34be6f4a2e/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs#L1453